volatility3/framework/plugins/pgdscan.py

```diff
@@ -0,0 +1,422 @@
+            layer = self.context.layers[layer.config["memory_layer"]]
+
+        # TODO: test and support other intel layers, either automatically
+        # detecting the likely type or allowing the user to provide it as
+        # a requirement option.
+        intel_class = intel.Intel32e
+
+        # get max layer address, this is used to validate possible PGDs as they
+        # cannot have pointers beyond the end of physical memory
+        maximum_address = layer.maximum_address
+
+        offsets = self.config.get("offset")
+        if offsets:
+            sections = [(offset, intel_class.page_size) for offset in offsets]
+        else:
+            sections = None
+
+        # store results of the scanning in a lookup so that the most frequent result
+        # can then be shown to the user.
+        khash_lookup = {}
+
+        # Run the scan
+        for pgd_offset, _pgd_data, khash in layer.scan(
+            self.context,
+            PageGlobalDirectoryScanner(maximum_address, intel_class=intel_class),
+            self._progress_callback,
+            sections=sections,
+        ):
+            if khash not in khash_lookup:
+                khash_lookup[khash] = []
+            khash_lookup[khash].append(pgd_offset)
+
+        # join is used a lot when building temp layers, this is simply
+        # here to make the code a little easier to read
+        join = interfaces.configuration.path_join
+
+        # find the most common khash, given that all user processes
+        # share the same kernel it is the most common khash that will
+        # locate the likely pgds
+
+        max_pgd_count = 0
+        most_common_khash = ""
+        for khash, pgds in khash_lookup.items():
+            if len(pgds) > max_pgd_count:
+                max_pgd_count = len(pgds)
+                most_common_khash = khash
+
+        for pgd_offset in khash_lookup[most_common_khash]:
+
+            # build a new layer for this likely pgd
+            temp_context = self.context.clone()
+            temp_layer_name = self.context.layers.free_layer_name("IntelLayer")
+            # temp_layer_name = "primary" # I would like to use the name primary but not sure how?
+            config_path = join("IntelHelper", temp_layer_name)
+            temp_context.config[join(config_path, "memory_layer")] = "memory_layer"
+            temp_context.config[join(config_path, "page_map_offset")] = pgd_offset
+            temp_layer = intel_class(
+                temp_context,
+                config_path=config_path,
+                name=temp_layer_name,
+            )
+            temp_context.add_layer(temp_layer)
+
+            config_fname = "-"
+            if self.config.get("save-configs"):
+                # TODO: Fix this. It seems like an ungly hack and must to the wrong way
+                # to make a new config with a new primary layer?
+                conf = {}
+                for key, value in dict(temp_layer.build_configuration()).items():
```

Comment:
That's how I would/have done it.  Definitely kinda of hacky, but I'm working on making the components of a config more reusable (by tagging their requirement type so it can be applied to "best guess" requirments of a similar type).

```
            new_config = {}
            config_dict = dict(primary.build_configuration())
            for entry in config_dict:
                # Volatility 1.2 support
                new_config["kernel.layer_name." + entry] = config_dict[entry]
                # Volatility <1.2 support
                new_config["primary." + entry] = config_dict[entry]
            json_str = json.dumps(new_config, sort_keys=True, indent=2)
```

Reply:
I mean if it's how you would have thought to do it that's got to be a compliment! :D I'll reword the TODO so it's worded more professionally and make a note to revisit it when you get time to add those config bits.